src/index.ts

```diff
@@ -27,15 +30,27 @@ const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
   return fail(new Error('' + err));
 };
 
-export const go = async <T, E extends Error>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T, E>> => {
-  // We need try/catch because `fn` might throw sync errors as well
+export const go = async <T, E extends Error>(
+  fn: () => Promise<T>,
+  options?: PromiseOptions
+): Promise<GoResult<T, E>> => {
   try {
-    if (typeof fn === 'function') {
-      return fn()
+    if (options?.retries) {
+      const optionsWithRetries = { ...options, retries: options.retries! };
+      return retryOperation(fn, optionsWithRetries)
+        .then(success)
+        .catch((err) => createGoError(err));
+    }
+
+    if (options?.timeoutMs) {
+      return promiseTimeout(options.timeoutMs, fn())
         .then(success)
         .catch((err) => createGoError(err));
     }
-    return fn.then(success).catch((err) => createGoError(err));
+
+    return fn()
+      .then(success)
+      .catch((err) => createGoError(err));
```

Comment:
Why treat this as 3 different cases? As far as I can tell you can always pass it through the `retry` function. With zero retries and no timeout if needed. I understand that the library has a bit of strange behavior, stopping retrying once the timeout is reached. But I think that can be handled differently. If we have both multiple retries and timeout and we run out of time, we can just run the whole thing again with one less retry. I haven't actually tried to write the code but I think it might work.